File: example-item-list/src/main/java/com/gurunars/item_list/example/AnimalItem.java
package com.gurunars.item_list.example;


import com.gurunars.item_list.Item;

public class AnimalItem implements Item {

    enum Type {
        MONKEY, TIGER, WOLF, LION
    }

    private Type type;
    private long id;
    private int version;

    public void update() {
        this.version++;
    }

    public AnimalItem(long id, Type type) {
        this.id = id;
        this.type = type;
    }

    @Override
    public Enum getType() {
        return type;
    }

    @Override
    public long getId() {
        return id;
    }

    @Override
    public boolean equals(Object obj) {
        if (! (obj instanceof AnimalItem)) {
            return false;
        }
        AnimalItem other = (AnimalItem) obj;
        return id == other.id && type == other.type && version == other.version;
    }

    @Override
    public String toString() {
        return "" + id + " @ " + version;
    }
}

File: examples/selectable-item-list/src/main/kotlin/com/gurunars/item_list/selectable_example/AnimalItem.kt
package com.gurunars.item_list.selectable_example

import com.gurunars.item_list.Item

internal data class AnimalItem(private val id: Long, val version: Int, private val type: AnimalItem.Type) : Item {

    override fun getType(): Enum<*> {
        return type
    }

    override fun getId(): Long {
        return id
    }

    override fun payloadsEqual(other: Item): Boolean {
        return other is AnimalItem && version == other.version
    }

    internal enum class Type {
        MONKEY, TIGER, WOLF, LION, EMPTY
    }

    override fun toString(): String {
        return "#$id{$type @ $version}"
    }

}
